File: leetcode/146-lru-cache/main.go
package main

/*
 */

/*
	initial thought:
	- use a hashtable to store the key & value to archieve put & get Time = O(1)
	- however, we want to prioritize the items by its recency.
	1. We should need an sorted array(sorted by recency), i.e. arr = arr[i]+arr[i+1]+target
	2. how to find the key value in the sorted array? what if we store the key & index such that we can loop up from the arr easier
	3. then the look up(from arr) complexity is O(1)
	- but wait, if we remove any item from the array, we need to update the hashtable, the set time complexity will be O(n)
	- what if we store a key & the pointer of the arr in the hashtable? such that it can remove itself from the arr?
	- but then we will take O(n) to iterate the array and remove the target item...
	- what if we use a doubly linked list? so when we remove an item,
	we can just look up the node pointer(doubly linked list) from the hashtable, then set node.prev.next = node.next and node.next.prev = node.prev
	- then the put & get Time = O(1)

	class approach:
	- use a hashtable to store the key and pointer of doubly linked list node
	- use doubly linked list to prioritize the item according to its recency
	Time if put & get O(1)
	Space							O(n) n: numbder of distinct key value
	18jan2019
*/

type ListNode struct {
	Key  int
	Val  int
	Prev *ListNode
	Next *ListNode
}

type LRUCache struct {
	Cap       int
	Cnt       int
	DumpHead  *ListNode
	DumpTail  *ListNode
	HashTable map[int]*ListNode // map is reference-typed, it always pass by pointer
}

func Constructor(capacity int) LRUCache {
	head := &ListNode{-1, -1, nil, nil}
	tail := &ListNode{-1, -1, nil, nil}
	head.Next = tail
	tail.Prev = head
	return LRUCache{capacity, 0, head, tail, make(map[int]*ListNode)}
}

func (this *LRUCache) Get(key int) int {
	if v, x := this.HashTable[key]; x {
		// move to tail
		this.moveToTail(v)
		return v.Val
	}
	return -1
}

func (this *LRUCache) Put(key int, value int) {
	if v, x := this.HashTable[key]; x {
		// move to tail
		v.Val = value
		this.moveToTail(v)
		this.HashTable[key] = v
		return
	}
	// add to tail
	node := &ListNode{key, value, nil, nil}
	this.addToTail(node)
	this.Cnt++
	// remove head if cnt > cap
	if this.Cnt > this.Cap {
		this.removeHead()
		this.Cnt--
	}
	this.HashTable[key] = node
}

func (this *LRUCache) addToTail(node *ListNode) {
	last := this.DumpTail.Prev
	last.Next = node
	node.Prev = last
	node.Next = this.DumpTail
	this.DumpTail.Prev = node
}

func (this *LRUCache) removeHead() {
	first := this.DumpHead.Next
	this.DumpHead.Next = first.Next
	first.Next.Prev = this.DumpHead
	delete(this.HashTable, first.Key)
}

func (this *LRUCache) moveToTail(node *ListNode) {
	node.Prev.Next = node.Next
	node.Next.Prev = node.Prev
	this.addToTail(node)
}

func main() {

}

File: leetcode/146-lru-cache/main.py
from collections import *
"""
    Doubly Linked List + Hashtable
    - similar to lc379, lc1429

    Time  put:O(1), get: O(1)
    124ms beats 96.42%
"""


class LLNode:
    def __init__(self, key=-1, val=-1):
        self.key = key
        self.val = val
        self.prev = None
        self.next = None


class LRUCache:

    def __init__(self, capacity: int):
        self.cap = capacity
        self.size = 0
        self.head = LLNode()
        self.tail = LLNode()
        self.head.next = self.tail
        self.tail.prev = self.head
        self.ht = {}

    def _removeNode(self, node):
        node.prev.next = node.next
        node.next.prev = node.prev

    def _addToTail(self, node):
        last = self.tail.prev
        last.next = node
        node.prev = last
        node.next = self.tail
        self.tail.prev = node

    def get(self, key: int) -> int:
        if key not in self.ht:
            return -1
        node = self.ht[key]
        self._removeNode(node)
        self._addToTail(node)
        return node.val

    def put(self, key: int, value: int) -> None:
        if key not in self.ht:
            node = LLNode(key, value)
            self._addToTail(node)
            self.ht[key] = node
            self.size += 1
        else:
            node = self.ht[key]
            node.val = value
            self._removeNode(node)
            self._addToTail(node)
        if self.size > self.cap:
            first = self.head.next
            self._removeNode(first)
            del self.ht[first.key]
            self.size -= 1


c = LRUCache(2)
c.put("c", 123)
c.put("a", 456)
c.put("b", 789)
print(c.map)
c.put("a", 123)
print(c.listHead)
c.get("b")
c.get("b")
c.get("b")
print(c.listHead)

"""
    2nd approach: use orderdict

    Time of put():    O(1)
    Time of get():    O(1)
    940 ms, faster than 6.85%
"""


class LRUCache(object):

    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self.capacity = capacity
        self.od = OrderedDict()

    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        if key in self.od:
            v = self.od[key]
            del self.od[key]
            self.od[key] = v
            return v
        return -1

    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: void
        """
        if key in self.od:
            del self.od[key]
            self.od[key] = value
        else:
            self.od[key] = value
            if len(self.od) > self.capacity:
                firstKey = None
                for key in self.od:
                    firstKey = key
                    break
                del self.od[firstKey]


c = LRUCache(2)
c.put("c", 123)
c.put("a", 456)
c.put("b", 789)
print(c.od)
c.put("a", 123)
print(c.od)
c.get("b")
c.get("b")
c.get("b")
print(c.od)
